foundation/org.eclipse.persistence.core/src/main/java/org/eclipse/persistence/internal/helper/ConcurrencyManager.java

```diff
@@ -46,26 +45,58 @@
+    private static final AtomicLong CONCURRENCY_MANAGER_ID = new AtomicLong(0);
 
     protected static boolean shouldTrackStack = PrivilegedAccessHelper.getSystemProperty(SystemProperties.RECORD_STACK_ON_LOCK) != null;
 
-    protected int numberOfReaders;
-    protected int depth;
-    protected int numberOfWritersWaiting;
+    protected AtomicInteger numberOfReaders;
+    protected AtomicInteger depth;
+    protected AtomicInteger numberOfWritersWaiting;
     protected volatile transient Thread activeThread;
 
     protected boolean lockedByMergeManager;
     protected Exception stack;
 
+    // Extended logging info fields
+    // Unique ID assigned each time when a new instance of a concurrency manager is created
+    private final long concurrencyManagerId = CONCURRENCY_MANAGER_ID.incrementAndGet();
+    // Creation date
+    private final Date concurrencyManagerCreationDate = new Date();
+    // In case if two threads are working on the exact same entity that leads to both threads wanting to release the same cache key
+    // there is tracking each increment of number of readers and their release.
+    private final AtomicLong totalNumberOfKeysAcquiredForReading = new AtomicLong(0);
+    // Same as totalNumberOfKeysAcquiredForReading but incremented each time the cache key is suffering to release cache key.
+    private final AtomicLong totalNumberOfKeysReleasedForReading = new AtomicLong(0);
+     // Total number of times the cache key caused a blow up because it suffered a release of cache key when the counter
+     // was set to 0. It should happen if an entity being shared by two threads.
+    private final AtomicLong totalNumberOfKeysReleasedForReadingBlewUpExceptionDueToCacheKeyHavingReachedCounterZero = new AtomicLong(0);
+
+    private static final Map<Thread, ConcurrencyManager> THREADS_TO_WAIT_ON_ACQUIRE_READ_LOCK = new ConcurrentHashMap<>();
+    private static final Map<Thread, ConcurrencyManager> THREADS_TO_WAIT_ON_ACQUIRE = new ConcurrentHashMap<>();
+    // Holds as a keys threads that needed to acquire one or more read locks on different cache keys.
+    private static final Map<Thread, ReadLockManager> READ_LOCK_MANAGERS = new ConcurrentHashMap<>();
+    private static final Set<Thread> THREADS_WAITING_TO_RELEASE_DEFERRED_LOCKS = ConcurrentHashMap.newKeySet();
+
+    /**
+     * In the places where we use this constant normally if a thread is stuck it is because it is doing object building.
+     * Blowing the threads ups is not that dangerous. It can be very dangerous for production if the dead lock ends up
+     * not being resolved because the productive business transactions will become cancelled if the application has a
+     * limited number of retries to for example process an MDB. However, the code spots where we use this constant are
+     * not as sensible as when the write lock manager is starving to run commit.
+     *
+     */
+    private static final Boolean ALLOW_INTERRUPTED_EXCEPTION_TO_BE_FIRED = true;
```

Comment:
this cannot be set to a different value - is it expected?

Reply:
It's transformed into JPA persistence property.